ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadState.java

```diff
@@ -0,0 +1,218 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.sync.worldstate;
+
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.eth.manager.task.EthTask;
+import tech.pegasys.pantheon.ethereum.worldstate.WorldStateStorage;
+import tech.pegasys.pantheon.ethereum.worldstate.WorldStateStorage.Updater;
+import tech.pegasys.pantheon.services.queue.TaskQueue;
+import tech.pegasys.pantheon.services.queue.TaskQueue.Task;
+import tech.pegasys.pantheon.util.ExceptionUtils;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+class WorldDownloadState {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final TaskQueue<NodeDataRequest> pendingRequests;
+  private final ArrayBlockingQueue<Task<NodeDataRequest>> requestsToPersist;
+  private final int maxOutstandingRequests;
+  private final Set<EthTask<?>> outstandingRequests =
+      Collections.newSetFromMap(new ConcurrentHashMap<>());
+  private final AtomicBoolean sendingRequests = new AtomicBoolean(false);
+  private final CompletableFuture<Void> future;
+  private final CompletableFuture<Void> tearDownFuture;
+  private boolean waitingForNewPeer = false;
+  private BytesValue rootNodeData;
+  private EthTask<?> persistenceTask;
+
+  public WorldDownloadState(
+      final TaskQueue<NodeDataRequest> pendingRequests,
+      final ArrayBlockingQueue<Task<NodeDataRequest>> requestsToPersist,
+      final int maxOutstandingRequests) {
+    this.pendingRequests = pendingRequests;
+    this.requestsToPersist = requestsToPersist;
+    this.maxOutstandingRequests = maxOutstandingRequests;
+    this.future = new CompletableFuture<>();
+    this.tearDownFuture = new CompletableFuture<>();
+    this.future.whenComplete(this::doCancel);
+    this.tearDownFuture.exceptionally(
+        error -> {
+          // Propagate cancellation back to our internal future.
+          if (error instanceof CancellationException) {
+            this.future.cancel(true);
+          }
+          return null;
+        });
+  }
+
+  private synchronized void doCancel(final Void result, final Throwable error) {
```

Comment:
Maybe rename this to `onShutdown` or `cleanup` or something since it always runs?

Reply:
Yep, went with cleanup.